python/grass/app/runtime.py

```diff
@@ -29,15 +29,34 @@ def get_grass_config_dir(major_version, minor_version, env):
     Determines path of GRASS GIS user configuration directory.
     """
     # The code is in sync with grass.app.runtime (but not the same).
+    env_dirname = "APPDATA" if WINDOWS else "HOME"
+    if env.get(env_dirname) is None:
+        raise KeyError(
+            f"The {env_dirname} variable is not set, ask your operating"
+            " system support"
+        )
+    if not os.path.exists(env.get(env_dirname)):
+        raise IsADirectoryError(
+            f"The {env_dirname} variable points to directory which does"
+            " not exist, ask your operating system support"
+        )
+
     if WINDOWS:
         config_dirname = f"GRASS{major_version}"
-        return os.path.join(env.get("APPDATA"), config_dirname)
+        config_path = os.path.join(env.get(env_dirname), config_dirname)
     elif MACOS:
-        version = f"{major_version}.{minor_version}"
-        return os.path.join(env.get("HOME"), "Library", "GRASS", version)
+        config_dirname = os.path.join("GRASS", f"{major_version}.{minor_version}")
+        config_path = os.path.join(env.get(env_dirname), "Library")
     else:
         config_dirname = f".grass{major_version}"
-        return os.path.join(env.get("HOME"), config_dirname)
+        config_path = os.path.join(env.get(env_dirname), config_dirname)
+
+    if env.get("GRASS_CONFIG_DIR"):
+        # override config_path if GRASS_CONFIG_DIR environmental
+        # variable is defined
+        config_path = os.path.join(env.get("GRASS_CONFIG_DIR"), config_dirname)
```

Comment:
This overrides, so it can be first and no tests of base env vars or existence of paths from them is needed. (Current code would raise if `HOME` does not exist even when `GRASS_CONFIG_DIR` is set.)